README.md

```diff
@@ -1,105 +1,273 @@
+___
+
+```python
+def insert_te(self, pos: int, length: int) -> int:
+        pos = pos % len(self.genome) # O(1)
+        active_tes = list(self.te_identities.items()) # O(k)
+        for id, te_range in active_tes: # O(k)
+            start, end = te_range # O(1)
+            if start <= pos < end: # O(1)
+                self.genome[start:end] = (end - start) * ["x"] # O(n + m)
+                self.te_identities.pop(id) # O(1)
+                continue
+            if start > pos: # O(1)
+                self.te_identities[id] = (start + length, end + length) # O(1)
+        self.te_count += 1 # O(1)
+        self.te_identities[self.te_count] = (pos, pos + length) # O(1)
+        self.genome = self.genome[:pos] + length * ["A"] + self.genome[pos:] # O(n + m)
+        return self.te_count
+```
+**insert_te complexity:** $O(k(n+m))$
+___
+```python
+  def copy_te(self, te: int, offset: int) -> int | None:
+        if self.te_identities[te]: # O(1)
+            te_range = self.te_identities[te] # O(1)
+            start, end = te_range # O(1)
+            copy_length = end - start # O(1)
+            copy_pos = start + offset # O(1)
+        return self.insert_te(copy_pos, copy_length)
+```
+**copy_te complexity:** $O(1)/O(k(n+m))$
+
+While most of the copy-function can be completed in constant time, it also inherits the complexity of the insert function as a final step.
+___
+```python
+def disable_te(self, te: int) -> None:
+        if self.te_identities[te]: # O(1)
+            te_range = self.te_identities.pop(te) # O(1)
+            start, end = te_range # O(1)
+            self.genome[start:end] = (end - start) * ['x'] # O(k + n)
+```
+**disable_te complexity:** $O(k+n)$
+___
+```python
+    def active_tes(self) -> list[int]:
+        return list(self.te_identities.keys())
+```
+**active_tes complexity:** $O(k)$
+___
+```python
+def __len__(self) -> int:
+        return len(self.genome)
+```
+**\_\_len\_\_ complexity:** $O(1)$
+___
+```python
+def __str__(self) -> str:
+        return "".join(self.genome)
+```
+**\_\_str\_\_ complexity:** $O(n)$
+___
+
+### Doubly-linked list
+
+```python
+def __init__(self, n: int):
+        self.head = Link(None, None, None) # O(1)
+        self.head.next = self.head # O(1)
+        self.head.prev = self.head # O(1)
+        self.length = n # O(1)
+        self.te_identities = {} # O(1)
+        self.te_count = 0 # O(1)
+        for _ in range(n): # O(n)
+            insert_after(self.head.prev, "-") # O(1)
+```
+**\_\_init\_\_ complexity:** $O(n)$
+___
+```python
+def insert_te(self, pos: int, length: int) -> int:
+        active_tes = list(self.te_identities.items()) # O(k)
+        for id, te_range in active_tes: # O(k)
+            start, end = te_range # O(1)
+            if start <= pos < end: # O(1)
+                self.disable_te(id)
+                continue
+            if start > pos: # O(1)
+                self.te_identities[id] = (start + length, end + length) # O(1)

```

Comment:
Hvorfor vil du egentlig kende deres position? Den har du vel ikke noget at bruge til. Hvis du i stedet lægger et TE ID i hver link, og for hver aktiv TE har en pointer til det første link, så kan du bevæge dig imellem positioner og TEer i konstant tid. Det vil reducere denne del til $O(1)$. Du skal finde positionen (det koster $O(n)$ under alle omstændigheder) men så kan du i konstant tid se om der ligger en aktiv TE der, og hvis der gør, kan du nuke den.